File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        postfix, prefix, ans = [1] * len(nums), [1] * len(nums), []
        
        for i in range(1, len(nums)):
            prefix[i] = prefix[i-1] * nums[i-1]
            
        for i in range(len(nums)-2, -1, -1):
            postfix[i] = postfix[i+1] * nums[i+1]
            
        print(prefix, postfix)
        for i in range(len(prefix)):
            ans.append(postfix[i] * prefix[i])
            
        return ans

File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int n = nums.size();
        vector<int> prefix_sum(n, 1);
        vector<int> postfix_sum(n, 1);
        vector<int> ans;
        /*
        [1, 1, 2, 6]
        [24, 12, 4, 1]
        
        */
        for (int i = 1; i < n; i++){
            prefix_sum[i] = prefix_sum[i-1] * nums[i-1];
        }

        for (int i = n-2; i >= 0; i--){
            postfix_sum[i] = postfix_sum[i+1] * nums[i+1];
        }

        for (int i = 0; i < n; i++){
            ans.push_back(prefix_sum.at(i) * postfix_sum.at(i));
        }

        return ans;
    }
};
